File: backend/scripts/test-mongodb.py
#!/usr/bin/env python3
import sys
import os
import re
import socket

"""
MongoDB Connection & Port Tester
Usage:
    python3 scripts/test-mongodb.py "mongodb://travel2d_travel2dubai:YOUR_PASSWORD@127.0.0.1:27017/travel2d_travel2dubai?authSource=travel2d_travel2dubai"
"""

def test_socket(host, port, timeout=3):
    print(f"[*] Checking TCP port reachability on {host}:{port}...")
    try:
        sock = socket.create_connection((host, port), timeout=timeout)
        sock.close()
        print(f"✅ Port {port} on {host} is reachable via TCP!")
        return True
    except Exception as e:
        print(f"❌ Port {port} on {host} is NOT reachable: {e}")
        return False

def main():
    uri = sys.argv[1] if len(sys.argv) > 1 else os.getenv("MONGODB_URI")
    if not uri:
        uri = "mongodb://travel2d_travel2dubai:travel2dubai@127.0.0.1:27017/travel2d_travel2dubai?authSource=travel2d_travel2dubai"

    if "YOUR_PASSWORD" in uri:
        print("\n⚠️ WARNING: URI contains placeholder 'YOUR_PASSWORD'. Please replace it with your actual password.\n")

    masked_uri = re.sub(r':([^:@]+)@', ':*****@', uri)
    print(f"[+] Testing URI: {masked_uri}")

    # Parse host & port
    match = re.search(r'@([^:/]+)(?::(\d+))?', uri)
    if not match:
        # standard fallback if no auth
        match = re.search(r'mongodb://([^:/]+)(?::(\d+))?', uri)

    host = match.group(1) if match else "127.0.0.1"
    port = int(match.group(2)) if match and match.group(2) else 27017

    port_ok = test_socket(host, port)

    try:
        import pymongo
        print("\n[*] pymongo is installed. Attempting full MongoDB authentication check...")
        client = pymongo.MongoClient(uri, serverSelectionTimeoutMS=5000)
        # Force server roundtrip
        info = client.server_info()
        print("✅ MongoDB connection and authentication SUCCESSFUL!")
        print(f"[+] Server MongoDB version: {info.get('version')}")

        # Extract db name
        db_match = re.search(r'27017/([^?]+)', uri)
        db_name = db_match.group(1) if db_match else "travel2d_travel2dubai"
        db = client[db_name]
        cols = db.list_collection_names()
        print(f"[+] Collections in '{db_name}': {cols}")
    except ImportError:
        print("\n[!] pymongo library is not installed locally. (TCP port check performed above)")
        print("    To run full authentication test with Python: pip install pymongo")
    except Exception as e:
        print(f"\n❌ MongoDB connection/auth failed: {e}")

if __name__ == "__main__":
    main()

File: backend/scripts/test-mongodb.js
const mongoose = require('mongoose');

/**
 * MongoDB Connection Test Script
 * Usage:
 *   node scripts/test-mongodb.js "mongodb://travel2d_travel2dubai:YOUR_PASSWORD@127.0.0.1:27017/travel2d_travel2dubai?authSource=travel2d_travel2dubai"
 *   OR set environment variable MONGODB_URI
 */

async function testConnection() {
  let uri = process.argv[2] || process.env.MONGODB_URI;

  if (!uri) {
    uri = 'mongodb://travel2d_travel2dubai:travel2dubai@127.0.0.1:27017/travel2d_travel2dubai?authSource=travel2d_travel2dubai';
  }

  // Check if user forgot to replace YOUR_PASSWORD
  if (uri.includes('YOUR_PASSWORD')) {
    console.warn('\n⚠️ WARNING: The URI contains "YOUR_PASSWORD". Replace "YOUR_PASSWORD" with the actual password before running.\n');
  }

  // Mask password for logging
  const maskedUri = uri.replace(/:([^:@]+)@/, ':*****@');
  console.log(`[+] Attempting to connect to MongoDB: ${maskedUri}`);

  try {
    await mongoose.connect(uri, {
      serverSelectionTimeoutMS: 5000, // Timeout after 5 seconds
      connectTimeoutMS: 5000,
    });

    console.log('✅ Connection SUCCESSFUL!');
    const db = mongoose.connection.db;
    console.log(`[+] Connected Database Name: "${db.databaseName}"`);

    // List collections to verify read access
    const collections = await db.listCollections().toArray();
    console.log(`[+] Collections count: ${collections.length}`);
    if (collections.length > 0) {
      console.log('[+] Existing collections:', collections.map(c => c.name).join(', '));
    } else {
      console.log('[+] Database currently has no collections.');
    }

    await mongoose.disconnect();
    console.log('[+] Connection closed cleanly.\n');
    process.exit(0);
  } catch (error) {
    console.error('\n❌ Connection FAILED!');
    console.error(`[!] Error Code: ${error.code || 'UNKNOWN'}`);
    console.error(`[!] Error Message: ${error.message}\n`);

    if (error.message.includes('ECONNREFUSED')) {
      console.error('💡 Recommendation: MongoDB service is not running on 127.0.0.1:27017 locally, or port 27017 is blocked.');
      console.error('   If MongoDB is on VPS (103.138.96.92), run this script directly on the VPS or use an SSH tunnel:');
      console.error('   ssh -L 27017:127.0.0.1:27017 root@103.138.96.92\n');
    } else if (error.message.includes('Authentication failed') || error.code === 18) {
      console.error('💡 Recommendation: The username or password is incorrect, or the authSource database is invalid.\n');
    }

    process.exit(1);
  }
}

testConnection();
